src/hooks/useSelector.ts

```diff
@@ -44,13 +44,13 @@ export function createSelectorHook(
       }
     }
 
-    const { store } = useReduxContext()!
+    const { store, getServerState } = useReduxContext()!
 
     const selectedState = useSyncExternalStoreWithSelector(
       store.subscribe,
       store.getState,
       // TODO Need a server-side snapshot here
-      store.getState,
+      getServerState || store.getState,
```

Comment:
I wonder - do you maybe have a good reference for the `getServerState` snapshot API~? I've tried searching through WG discussions but I came empty-handed. I recall reading about this but would love to refresh myself on why this is needed with `uSES` and how React avoid the need for it with its primitives

Reply:
> I've tried searching through WG discussions but I came empty-handed.

https://github.com/facebook/react/issues/22361 might help. Note that my initial post made some false assumptions that were corrected.

> I recall reading about this but would love to refresh myself on why this is needed with uSES

By the time a particular component is hydrated, the store might've already been mutated. Simplest example I encountered was something like

```jsx
<MutatesStoreInAnEffect />
<ReadsStore />
<React.Suspense>
  <ReadsStore />
</React.Suspense>
```

Assuming no component suspends, the server would render the same value for the store in every `ReadsStore`. But if you hydrate this tree then the second `ReadsStore` might be hydrated after the effects of `MutatesStoreInAnEffect` are run (due to selective hydration). The first `ReadsStore` would hydrate fine because we still have the initial state.

So we need an API to tell React with what state the markup was rendered on the server.

> how React avoid the need for it with its primitives

Not sure I understand this statement correctly but React does need `getServerSnapshot` in every version. It needs it even more in React 18. In React 17 you'd be safe against hydration mismatch if your store is never mutated during render. In React 18 you can even get hydration mismatches if your store is mutated in effects.

Reply:
Thank you for the link and the explanation. This is what I was roughly expecting. I didn't clearly understand how the client value is handled though, and according to Andrew, the "mismatched" component is just re-rendered on top of the hydrated component that was hydrated based on the server snapshot.

> Not sure I understand this statement correctly but React does need getServerSnapshot in every version. It needs it even more in React 18. In React 17 you'd be safe against hydration mismatch if your store is never mutated during render. In React 18 you can even get hydration mismatches if your store is mutated in effects.

I'm wondering about the situation when no external store is involved. Just replace `MutatesStoreInAnEffect` with `UpdatesStateInAnEffect` and `ReadsStore` with `ReadsPropOrContextValueThatComesFromReactState` in your example. This is basically the very same situation as the described by your, but `useState` doesn't support any `getServerSnapshot` method:
https://github.com/facebook/react/blob/5cccacd131242bdea2c2fe4b33fac50d2e3132b4/packages/react-reconciler/src/ReactFiberHooks.new.js#L2592-L2594

So I wonder, how the implementation is different for both and how React can handle this better for a plain state value?

Reply:
One thing I've learned from my past attempts at debugging React code is that I'm not that good at reasoning about such sublime code 🤣 

Maybe @acdlite would be able to shed some light on this.

Reply:
The reason useState "just works" is because you can't update a component until after it has already hydrated